src/i18n/i18n.js

```diff
@@ -0,0 +1,508 @@
+     * translations from these assets as the assets are loaded and it will also automatically unload translations if the assets get removed or unloaded at runtime.
+     */
+    var I18n = function (app) {
+        pc.events.attach(this);
+
+        this.locale = DEFAULT_LOCALE;
+        this._translations = {};
+        this._availableLangs = {};
+        this._app = app;
+        this._assets = [];
+        this._parser = new pc.I18nParser();
+    };
+
+
+    /**
+     * @private
+     * @function
+     * @name pc.I18n#getText
+     * @description Returns the translation for the specified key and locale. If the locale is not specified
+     * it will use the current locale.
+     * @param {String} key The localization key
+     * @param {String} [locale] The desired locale.
+     * @returns {String} The translated text. If no translations are found at all for the locale then it will return
+     * the en-US translation. If no translation exists for that key then it will return the localization key.
+     * @example
+     * var localized = this.app.i18n.getText('localization-key');
+     * var localizedFrench = this.app.i18n.getText('localization-key', 'fr-FR');
+     */
+    I18n.prototype.getText = function (key, locale) {
+        var result = key;
+
+        var lang;
+        if (!locale) {
+            locale = this._locale;
+            lang = this._lang;
+        }
+
+        var translations = this._translations[locale];
+        if (!translations) {
+            if (!lang) {
+                lang = getLang(locale);
+            }
+
+            locale = this._findFallbackLocale(lang);
+            translations = this._translations[locale];
+        }
+
+        if (translations) {
+
+            // #ifdef DEBUG
+            if (Array.isArray(translations[key])) {
+                console.error('Invalid call to getText for key ' + key + ' and locale ' + locale + '. Use getPluralText instead for this key.');
+                return result;
+            }
+            // #endif
+
+            result = translations[key] || key;
+        }
+
+        return result;
+    };
+
+    /**
+     * @private
+     * @function
+     * @name pc.I18n#getPluralText
+     * @description Returns the pluralized translation for the specified key, number n and locale. If the locale is not specified
+     * it will use the current locale.
+     * @param {String} key The localization key
+     * @param {Nubmer} n The number used to determine which plural form to use. E.g. for the phrase "5 Apples" n equals 5.
+     * @param {String} [locale] The desired locale.
+     * @returns {String} The translated text. If no translations are found at all for the locale then it will return
+     * the en-US translation. If no translation exists for that key then it will return the localization key.
+     * @example
+     * var localized = this.app.i18n.getPluralText('5 apples', 5);
```

Comment:
slightly odd examples as you have the specific number in the KEY

Reply:
yeah I guess so just wanted to really point out what that '5' is in the example...

Reply:
I think the example shouldn't include the number, it does make it confusing.